pkg/resource/api.go

```diff
@@ -63,41 +67,119 @@ func (a *APIManagedConnectionPropagator) PropagateConnection(ctx context.Context
-	meta.AddAnnotations(mgcs, map[string]string{k: v})
+	k := strings.Join([]string{AnnotationKeyPropagateToPrefix, string(to.GetUID())}, AnnotationDelimiter)
+	v := strings.Join([]string{to.GetNamespace(), to.GetName()}, AnnotationDelimiter)
+	meta.AddAnnotations(from, map[string]string{k: v})
+
+	return errors.Wrap(a.client.Update(ctx, from), errUpdateSecret)
+}
+
+// An APIPatchingApplicator applies changes to an object by either creating or
+// patching it in a Kubernetes API server.
+type APIPatchingApplicator struct {
+	client client.Client
+}
+
+// NewAPIPatchingApplicator returns an Applicator that applies changes to an
+// object by either creating or patching it in a Kubernetes API server.
+func NewAPIPatchingApplicator(c client.Client) *APIPatchingApplicator {
+	return &APIPatchingApplicator{client: c}
+}
+
+// Apply changes to the supplied object. The object will be created if it does
+// not exist, or patched if it does.
+func (a *APIPatchingApplicator) Apply(ctx context.Context, o runtime.Object, ao ...ApplyOption) error {
+	m, ok := o.(metav1.Object)
+	if !ok {
+		return errors.New("cannot access object metadata")
+	}
+
+	desired := o.DeepCopyObject()
+
+	err := a.client.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, o)
+	if kerrors.IsNotFound(err) {
+		// TODO(negz): Apply ApplyOptions here too?
+		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
+	}
+	if err != nil {
+		return errors.Wrap(err, "cannot get object")
+	}
+
+	for _, fn := range ao {
+		if err := fn(ctx, o, desired); err != nil {
+			return err
+		}
+	}
+
+	return errors.Wrap(a.client.Patch(ctx, o, &patch{desired}), "cannot patch object")
+}
+
+type patch struct{ from runtime.Object }
+
+func (p *patch) Type() types.PatchType                 { return types.MergePatchType }
+func (p *patch) Data(_ runtime.Object) ([]byte, error) { return json.Marshal(p.from) }
+
+// An APIUpdatingApplicator applies changes to an object by either creating or
+// updating it in a Kubernetes API server.
+type APIUpdatingApplicator struct {
+	client client.Client
+}
+
+// NewAPIUpdatingApplicator returns an Applicator that applies changes to an
+// object by either creating or updating it in a Kubernetes API server.
+func NewAPIUpdatingApplicator(c client.Client) *APIUpdatingApplicator {
+	return &APIUpdatingApplicator{client: c}
+}
+
+// Apply changes to the supplied object. The object will be created if it does
+// not exist, or updated if it does.
+func (a *APIUpdatingApplicator) Apply(ctx context.Context, o runtime.Object, ao ...ApplyOption) error {
+	m, ok := o.(metav1.Object)
+	if !ok {
+		return errors.New("cannot access object metadata")
+	}
+
+	current := o.DeepCopyObject()
+
+	err := a.client.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, current)
+	if kerrors.IsNotFound(err) {
+		// TODO(negz): Apply ApplyOptions here too?
```

Comment:
I don't have a use case but I'd expect `ApplyOption`s that I supply to be called for all successful calls, which includes `Create`.

Reply:
Me too. This is used in a few places though so I don't want to risk altering the default logic too much in this PR, just in case.